torchtune/modules/vision_transformer.py

```diff
@@ -0,0 +1,334 @@
+
+        # patch embeddings (tokens)
+        # out: (bsz, embed_dim, patch_grid_size, patch_grid_size)
+        x = self.conv(images)
+
+        # out: (bsz, patch_grid_size**2, embed_dim)
+        x = x.flatten(start_dim=2).permute(0, 2, 1)
+
+        _, n_tokens, embed_dim = x.shape
+
+        # pre_tile_pos_embed
+        if self.pre_tile_pos_embed:
+            x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
+            x = self.pre_tile_pos_embed(x, aspect_ratio)
+
+        # apply cls token
+        x = x.reshape(bsz * n_tiles, n_tokens, embed_dim)
+        x = self.cls_token_embedding(x)
+        n_tokens += 1
+
+        # token_pos_embedding
+        x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
+        x = self.token_pos_embedding(x, aspect_ratio)
+
+        # norm
+        x = self.ln_pre(x)
+
+        # transformer with optional hidden layer outputs
+        x = x.view(bsz, n_tiles * n_tokens, embed_dim)
+        for layer_idx, transformer_layer in enumerate(self.transformer_layers):
+            if self.indices_return_hidden and layer_idx in self.indices_return_hidden:
+                hidden_states.append(x)
+            x = transformer_layer(x)
+
+        # norm
+        x = self.ln_post(x)
+
+        # reshape outputs
+        x = x.reshape(bsz, n_tiles, n_tokens, embed_dim)
+
+        if self.indices_return_hidden:
+            hidden_states = torch.stack(hidden_states, dim=-1)
+            hidden_states = hidden_states.reshape(
+                bsz, n_tiles, n_tokens, embed_dim, len(self.indices_return_hidden)
+            )
+        else:
+            hidden_states = torch.empty(0)  # dummy tensor
+
+        # post_tile_pos_embed
+        if self.post_tile_pos_embed:
+            x = self.post_tile_pos_embed(x, aspect_ratio)
+
+        # cls token projection
+        if self.cls_projection:
+            x = x.reshape(bsz * n_tiles, n_tokens, embed_dim)
+            x = self.cls_projection(x)
+            x = x.reshape(
+                bsz, n_tiles, 1, -1
+            )  # (bsz, n_tiles, num_tokens, cls_output_dim)
+
+        if self.indices_return_hidden:
+            return x, hidden_states
+        else:
+            return x
+
+
+class CLSEmbedding(nn.Module):
+    """
+    Adds a CLS token to every tile in an image.
+
+    Notice that tile is different from patch (token). An image is divided into tiles during pre-processing,
+    and patches are the outcome of the convolution in the ViT applied to each tile.
+
+    Args:
+        embed_dim (int): The dimensionality of each patch embedding.
+    Returns:
+        torch.Tensor: The input tensor with added CLS tokens.
+    """
+
+    def __init__(self, embed_dim: int) -> None:
+        super().__init__()
+
+        scale = embed_dim**-0.5
+        self.cls_embedding = nn.Parameter(scale * torch.randn(embed_dim))
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+
+        cls_emb = self.cls_embedding.to(x.dtype)
```

Comment:
Is there a specific reason we have to do this cast every time? My assumption would be that this is already initialized in the same dtype as all the other model components

Reply:
good catch! We dont do it for other layers either, and i dont see it in torchmultimodal. I will remove it.